services/repository-managers/src/main/java/org/sagebionetworks/repo/manager/DockerManagerImpl.java

```diff
@@ -0,0 +1,241 @@
+import org.sagebionetworks.repo.model.docker.DockerCommit;
+import org.sagebionetworks.repo.model.docker.DockerRegistryEvent;
+import org.sagebionetworks.repo.model.docker.DockerRegistryEventList;
+import org.sagebionetworks.repo.model.docker.DockerRepository;
+import org.sagebionetworks.repo.model.docker.RegistryEventAction;
+import org.sagebionetworks.repo.model.jdo.KeyFactory;
+import org.sagebionetworks.repo.model.principal.AliasType;
+import org.sagebionetworks.repo.model.principal.PrincipalAlias;
+import org.sagebionetworks.repo.model.util.DockerNameUtil;
+import org.sagebionetworks.repo.transactions.WriteTransaction;
+import org.sagebionetworks.repo.web.NotFoundException;
+import org.springframework.beans.factory.annotation.Autowired;
+
+
+
+public class DockerManagerImpl implements DockerManager {
+	@Autowired
+	private NodeDAO nodeDAO;
+
+	@Autowired
+	IdGenerator idGenerator;
+
+	@Autowired
+	UserManager userManager;
+
+	@Autowired
+	EntityManager entityManager;
+
+	@Autowired
+	AuthorizationManager authorizationManager;
+
+	/**
+	 * Answer Docker Registry authorization request.
+	 * 
+	 * @param userId
+	 * @param service
+	 * @param scope
+	 * @return
+	 */
+	@Override
+	public DockerAuthorizationToken authorizeDockerAccess(String userName, UserInfo userInfo, String service, String scope) {
+		String[] scopeParts = scope.split(":");
+		if (scopeParts.length!=3) throw new RuntimeException("Expected 3 parts but found "+scopeParts.length);
+		String type = scopeParts[0]; // type='repository'
+		String repositoryPath = scopeParts[1]; // i.e. the 'path'
+		String accessTypes = scopeParts[2]; // e.g. push, pull
+
+		List<String> permittedAccessTypes = getPermittedAccessTypes( userName,  userInfo,  service,  type, repositoryPath, accessTypes);
+		// now construct the auth response and return it
+		long now = System.currentTimeMillis();
+		String uuid = UUID.randomUUID().toString();
+
+		String token = DockerTokenUtil.createToken(userName, type, service, repositoryPath, 
+				permittedAccessTypes, now, uuid);
+		DockerAuthorizationToken result = new DockerAuthorizationToken();
+		result.setToken(token);
+		return result;
+
+	}
+
+	public List<String> getPermittedAccessTypes(String userName, UserInfo userInfo, 
+			String service, String type, String repositoryPath, String accessTypes) {
+
+		List<String> permittedAccessTypes = new ArrayList<String>();
+
+		String parentId = validParentProjectId(repositoryPath);
+
+		if (parentId==null) return permittedAccessTypes;
+
+		String entityName = service+DockerNameUtil.REPO_NAME_PATH_SEP+repositoryPath;
+		EntityHeader entityHeader = null;
+		try {
+			entityHeader = nodeDAO.getEntityHeaderByChildName(parentId, entityName);
+		} catch (NotFoundException nfe) {
+			entityHeader = null;
+		}
+		String existingDockerRepoId = null;
+		if (entityHeader!=null && EntityType.valueOf(entityHeader.getType()) == EntityType.dockerrepo) {
+			existingDockerRepoId = entityHeader.getId();
```

Comment:
if entity exists and is not a docker repo, simply return from the function with an empty list